// Copyright 2021 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package wfa.panelmatch.client.batchlookup;

option java_package = "org.wfanet.panelmatch.client.batchlookup";
option java_multiple_files = true;

// An argument in every Request proto. It contains various knobs for tuning the
// system.
message ObliviousQueryParameters {}

// A request to get a public/private key pair
message GenerateKeysRequest {
  ObliviousQueryParameters parameters = 1;
}

message GenerateKeysResponse {
  bytes public_key = 1;
  bytes private_key = 2;
}

// Used in the request to encrypt a set of queries
message UnencryptedQuery {
  int64 shard_id = 1;
  int64 query_id = 2;
  int64 bucket_id = 3;
}

// A request to encrypt a set of queries
message EncryptQueriesRequest {
  ObliviousQueryParameters parameters = 1;
  bytes public_key = 2;
  bytes private_key = 3;
  repeated UnencryptedQuery unencrypted_query = 4;
}

// Used in the response to encrypt a set of queries
message EncryptedQuery {
  int64 shard_id = 1;
  int64 query_id = 2;
}

message EncryptQueriesResponse {
  repeated bytes ciphertexts = 1;
  repeated EncryptedQuery encrypted_query = 2;
}

// A request to remove RLWE-encryption from a set of results.
message DecryptQueriesRequest {
  ObliviousQueryParameters parameters = 1;
  bytes public_key = 2;
  bytes private_key = 3;
  repeated bytes encrypted_query_results = 4;
}

// This response contains queries with RLWE-encryption removed. The query
// results are encrypted with another encryption scheme separately which is not
// removed during this step.
message DecryptQueriesResponse {
  repeated bytes decrypted_query_results = 1;
}
